# CREDIT CARD FRAUD DETECTION

1) Build a machine learning model to identify fraudulent credit card
transactions.
2) Preprocess and normalize the transaction data, handle class
imbalance issues, and split the dataset into training and testing sets.
3) Train a classification algorithm, such as logistic regression or random
forests, to classify transactions as fraudulent or genuine.
Evaluate the model's performance using metrics like precision, recall, 
and F1-score, and consider techniques like oversampling or
undersampling for improving results.

### Datasets link: https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud

In [36]:
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn

# warnings
import warnings
warnings.filterwarnings('ignore')

# split data
from sklearn.model_selection import train_test_split

# Linear Regression
from sklearn.linear_model import LinearRegression, LogisticRegression

# Polynomial features
from sklearn.preprocessing import PolynomialFeatures

# calculate accuracy
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score, accuracy_score

In [3]:
df = pd.read_csv("creditcard.csv")
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [6]:
df["Class"].value_counts()

0    284315
1       492
Name: Class, dtype: int64

#### Number of legit transactions are: 284315 & illegal transactions are 492

number of legit transactions >> number of illegal transactions

In [7]:
legit_df = df[df["Class"] == 0]
illegal_df = df[df["Class"] == 1]

In [8]:
legit_df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [9]:
illegal_df.head()

Time        V1        V2        V3        V4        V5        V6  \
541    406.0 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
623    472.0 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
4920  4462.0 -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788   
6108  6986.0 -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536   
6329  7519.0  1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746   

            V7        V8        V9  ...       V21       V22       V23  \
541  -2.537387  1.391657 -2.770089  ...  0.517232 -0.035049 -0.465211   
623   0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
4920  0.562320 -0.399147 -0.238253  ... -0.294166 -0.932391  0.172726   
6108 -3.496197 -0.248778 -0.247768  ...  0.573574  0.176968 -0.436207   
6329  1.713445 -0.496358 -1.282858  ... -0.379068 -0.704181 -0.656805   

           V24       V25       V26       V27       V28  Amount  Class  
541   0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
623  -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
4920 -0.087330 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  
6108 -0.053502  0.252405 -0.657488 -0.827136  0.849573   59.00      1  
6329 -1.632653  1.488901  0.566797 -0.010016  0.146793    1.00      1  

[5 rows x 31 columns]

In [10]:
df.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

In [13]:
legit_rnd_df = legit_df.sample(1500)
legit_rnd_df.shape

(1500, 31)

In [14]:
new_df = pd.concat([legit_rnd_df, illegal_df])

In [19]:
new_df.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      92897.854000 -0.062629  0.038120 -0.001188  0.046150 -0.043564   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.093946  0.007970 -0.010879  0.045518  ... -0.016980  0.004688   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.006511 -0.034031  0.004266  0.003101  0.013215 -0.010941  0.001087   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       93.073780  
1      122.211321  

[2 rows x 30 columns]

### Data splting, training, testing

In [23]:
x = new_df.iloc[:, :30]
x.head()

Time        V1        V2        V3        V4        V5        V6  \
158090  110914.0 -3.064917  1.147388 -1.309701 -0.028826  0.844547 -2.001638   
102396   68183.0  0.624728 -1.339744  0.920933  0.445550 -1.372452  0.411138   
88289    62074.0 -1.302580  0.996009  1.475091  0.977379 -1.099666  0.396492   
114084   73329.0 -0.721877  0.941507  1.890013 -0.502963  0.225044 -0.998323   
55161    46812.0 -3.245073  2.561551  1.037907 -2.319756  0.308523 -0.535802   

              V7        V8        V9  ...       V20       V21       V22  \
158090  0.084812  0.853600  0.379904  ... -0.221701  0.172953 -0.042236   
102396 -0.579884  0.334320  1.381419  ...  0.247969 -0.024975 -0.376267   
88289  -0.770872  1.034844  0.329604  ... -0.139345  0.016721  0.088977   
114084  0.929954 -0.296577 -0.143235  ...  0.062384 -0.317988 -0.788327   
55161   1.653836 -1.145764  3.110089  ...  1.877958 -0.560651  0.010100   

             V23       V24       V25       V26       V27       V28  Amount  
158090 -0.927399 -0.249266  1.049719 -0.148464 -0.632463 -0.268672    1.00  
102396 -0.017072  0.139562 -0.198319  0.960646 -0.056614  0.050300  254.30  
88289  -0.105960 -0.029243  0.054890 -0.348405 -0.332181 -0.348745   12.99  
114084 -0.191954  0.356560  0.303070 -0.003537 -0.123485 -0.197014    2.68  
55161  -0.125529  0.198649 -0.020393  0.594722 -0.050623 -0.363087    3.84  

[5 rows x 30 columns]

In [26]:
y = new_df.iloc[: , 30]
y.head()

158090    0
102396    0
88289     0
114084    0
55161     0
Name: Class, dtype: int64

In [29]:
x = x.values
y = y.values

In [42]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.1, random_state = 5)

In [43]:
reg = LogisticRegression()

In [44]:
reg.fit(x_train, y_train)

LogisticRegression()

In [45]:
y_pred = reg.predict(x_test)

In [46]:
accuracy_score(y_pred, y_test)

0.975

### So, accuracy of predicting score is 97.5%